R/select_nse.R

```diff
@@ -0,0 +1,512 @@
+# (starting from the lower one) until the global environment. If we get its
+# value but it errors because the function doesn't exist then it means that
+# it is a select helper that we grab from the error message.
+
+.select_symbol <- function(data, x, ignore_case, regex, verbose) {
+  try_eval <- try(eval(x), silent = TRUE)
+  x_dep <- insight::safe_deparse(x)
+  is_select_helper <- FALSE
+  out <- NULL
+
+  if (x_dep %in% colnames(data)) {
+    matches <- match(x_dep, colnames(data))
+    out <- matches[!is.na(matches)]
+  } else if (isTRUE(ignore_case)) {
+    matches <- match(toupper(x_dep), toupper(colnames(data)))
+    out <- matches[!is.na(matches)]
+  } else {
+    new_expr <- tryCatch(
+      .dynGet(x, inherits = FALSE, minframe = 0L),
+      error = function(e) {
+        # if starts_with() et al. don't exist
+        fn <- insight::safe_deparse(e$call)
+
+        # if starts_with() et al. come from tidyselect but need to be used in
+        # a select environment, then the error doesn't have the same structure.
+        if (is.null(fn) &&
+          grepl("must be used within a", e$message, fixed = TRUE)) {
+          trace <- lapply(e$trace$call, function(x) {
+            tmp <- insight::safe_deparse(x)
+            if (grepl(paste0("^", .regex_select_helper()), tmp)) {
+              tmp
+            }
+          })
+          fn <- Filter(Negate(is.null), trace)[1]
+        }
+        # if we actually obtain the select helper call, return it, else return
+        # what we already had
+        if (length(fn) > 0 && grepl(.regex_select_helper(), fn)) {
+          is_select_helper <<- TRUE
+          return(fn)
+        } else {
+          NULL
+        }
+      }
+    )
+
+    if (is_select_helper) {
+      new_expr <- str2lang(unlist(new_expr))
+      out <- .eval_expr(
+        new_expr,
+        data = data,
+        ignore_case = ignore_case,
+        regex = regex,
+        verbose = verbose
+      )
+    } else if (length(new_expr) == 1L && is.function(new_expr)) {
+      out <- which(vapply(data, new_expr, FUN.VALUE = logical(1L)))
+    } else {
+      out <- unlist(lapply(
+        new_expr,
+        .eval_expr,
+        data = data,
+        ignore_case = ignore_case,
+        regex = regex,
+        verbose = verbose
+      ))
+    }
+  }
+
+  # sometimes an object that needs to be evaluated has the same name as a
+  # function (e.g colnames). Vector of names have the priority on functions
+  # so function evaluation is delayed at the max.
+  if (is.null(out) && is.function(try_eval)) {
+    cols <- names(data)
+    out <- which(vapply(data, x, FUN.VALUE = logical(1L)))
+  }
+
+  out
+}
+
+# Dispatch expressions to various select helpers according to the function call.
+
+.eval_call <- function(data, x, ignore_case, regex, verbose) {
+  type <- as.character(x[[1]])
+  if (length(type) > 1L) {
+    # This helps when pkg::fn is used in a select helper
```

Comment:
Can you include an example here of such usage? It's not clear to me what usage is being referred here.

Reply:
This was something specific to `poorman` (see [this issue](https://github.com/nathaneastwood/poorman/issues/105)), I'll remove it